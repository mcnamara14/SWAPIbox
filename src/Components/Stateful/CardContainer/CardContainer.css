.header {
  width: 80%;
  display: inline-block; }

.cards {
  background: url('./bg.jpg');
  padding-top: 40px; }

.cards div {
  display: flex;
  width: 80%;
  margin: 0 auto;
  flex-wrap: wrap;
  justify-content: space-between; }
