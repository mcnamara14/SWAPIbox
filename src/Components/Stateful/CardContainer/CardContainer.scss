.header {
  width: 80%;
  display: inline-block;
}

.cards {
  // background-color: #282046;
  background: url('./bg.jpg');
  padding-top: 40px;
}

.cards div {
  display: flex;
  width: 80%;
  margin: 0 auto;
  flex-wrap: wrap;
  justify-content: space-between;
}

.cardsWrapper {
  margin-bottom: 80px;
}

