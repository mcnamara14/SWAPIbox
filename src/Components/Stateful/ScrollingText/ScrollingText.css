.scrollingTextContainer {
  background-repeat: repeat !important;
  background: url("./bg.jpg");
  display: block;
  height: 100vh;
  padding-top: 40px; }

#titles {
  bottom: 0;
  font-size: 350%;
  font-weight: bold;
  height: 50em;
  left: 50%;
  margin-left: -9em;
  overflow: hidden;
  position: absolute;
  text-align: justify;
  transform-origin: 50% 100%;
  transform: perspective(300px) rotateX(25deg);
  width: 18em; }

span {
  color: white;
  margin-bottom: 20px; }

.scrollYear {
  color: white;
  font-size: 3rem; }

#titles:after {
  bottom: 0%;
  content: ' ';
  left: 0;
  pointer-events: none;
  position: absolute;
  right: 0;
  top: 0; }

#titlecontent {
  position: absolute;
  top: 100%;
  animation: scroll 80s linear 4s infinite;
  color: #d39f00; }

@keyframes scroll {
  0% {
    top: 100%; }
  100% {
    top: -170%; } }
