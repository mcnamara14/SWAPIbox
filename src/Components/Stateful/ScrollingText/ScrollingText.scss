$font-stack:    Helvetica, sans-serif;
$primary-color: #333;

.scrollingText {
  width: 20%;
  float: left;
  background: #eee;
  height: 100vh;
}

