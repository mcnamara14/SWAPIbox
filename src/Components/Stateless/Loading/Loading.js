import React from 'react';
import './Loading.css';
import loadingSound from './loading.wav';

const Header = (props) => {
  
  return (
    <div className="loading">
      <audio src={loadingSound} autoPlay loop></audio>
      <img src={require('./loading.gif')} height="400" alt={'Loading gif'}/>
    </div>
  );
}

export default Header;
