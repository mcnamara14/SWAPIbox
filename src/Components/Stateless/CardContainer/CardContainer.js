import React from 'react';
import Card from '../../Stateless/Card/Card';
import './CardContainer.css';

const CardContainer = ({ data, filter }) => {

    return (
      <article className="CardContainer">
        <Card data={data} filter={filter} />
      </article>
    )
}

export default CardContainer;
