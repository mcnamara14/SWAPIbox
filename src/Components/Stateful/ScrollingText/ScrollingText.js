import React, { Component } from 'react';
import './ScrollingText.css';
import themeSong from './imperial-march.mp3';

class ScrollingText extends Component {
  constructor() {
    super();
    
    this.state = {
      crawl: '', 
      title: '',
      date: ''
    }
  }
    
  componentDidMount = () => {
    fetch('https://swapi.co/api/films')
      .then(response => response.json())
      .then(response => this.returnRandomCrawl(response.results))
  }

  returnRandomCrawl = (films) => {
    const filmsLength = films.length;
    const randomFilm = Math.floor(Math.random() * filmsLength) + 0;
    const crawl = films[randomFilm].opening_crawl;
    const title = films[randomFilm].title;
    const date = films[randomFilm].release_date;

    this.setState({ crawl, title, date })
  }

  render() {
    return (
      <section className="scrollingTextContainer">
      <div id="titles">
        <div id="titlecontent">
        <audio src={themeSong} autoPlay loop></audio>
        <p>{this.state.crawl}</p>
        <p><span>{this.state.title}</span></p>
        <p className="scrollYear"><span>{this.state.date}</span></p>
    
      </div>
    </div>
    </section>

    );
  };
}

export default ScrollingText;
