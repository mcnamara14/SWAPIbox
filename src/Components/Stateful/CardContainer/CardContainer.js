import React, { Component } from 'react';
import Card from '../../Stateless/Card/Card';
import Buttons from '../../Stateless/Buttons/Buttons';
import DataCleaner from '../../../DataCleaner/DataCleaner';
import Loading from '../../Stateless/Loading/Loading';
import AddFavorites from '../../Stateless/AddFavorites/AddFavorites';
import ScrollingText from '../../Stateful/ScrollingText/ScrollingText';
import './CardContainer.css';

const dataCleaner = new DataCleaner();

class CardContainer extends Component {
  constructor() {
    super();

    this.state = {
      filter: '',
      data: {},
      renderState: 'crawl',
      favorites: [],
      favoriteCount: 0
    };

    this.displayFavorites = false;
    this.setData = this.setData.bind(this);
  }

    setData = async (filter) => {
      this.setState({ renderState: 'loading' });
      const data = await dataCleaner.fetchData(filter);
      this.setState({ data, renderState: 'cards' });
    }

    findCard = (id) => {
      const selectedCard = this.state.data.find(data => data.id === id);
      const duplicateFavorite = this.state.favorites
        .find(favorite => favorite.id === id);

      if (duplicateFavorite) {
        this.removeFavorite(selectedCard);
      } else {
        this.addFavorite(selectedCard);
      }
    }

    removeFavorite = (selectedCard) => {
      let updatedFavoriteCount = this.state.favoriteCount;
      let favoriteCount = updatedFavoriteCount -= 1;
      const favorites = this.state.favorites;
      const index = favorites
        .findIndex(favorite => favorite.id === selectedCard.id);

      favorites.splice(index, 1);

      this.setState({ favorites, favoriteCount });
    }

    addFavorite = (selectedCard) => {
      let updatedFavoriteCount = this.state.favoriteCount;
      let favoriteCount = updatedFavoriteCount += 1;
      const favorites = [...this.state.favorites, selectedCard];

      this.setState({ favorites, favoriteCount });
    };

    toggleDisplayFavorites = () => {
      this.setState({ data: this.state.favorites });
    }

    toggleAddFavorites = () => {
      this.setState({ renderState: 'addFavorites' })
    }

    getCards = () => {
      const cards = this.state.data.map((eachData, index) => {
        const selected = this.state.favorites
          .find(favorite => favorite.id === eachData.id);

        return <Card data={eachData} key={index} 
          findCard={this.findCard} selected={selected} />;
      });

      return cards;
    }

    toggleRender = () => {
      switch (this.state.renderState) {
        case 'crawl':
          return <ScrollingText />;
        case 'cards':
          return <section className="cards">
            <div>
              {this.getCards()}
            </div>
          </section>;
        case 'loading':
          return <div><Loading /></div>;
        case 'addFavorites': 
          return <AddFavorites />
      }
    }

    render() {
      return (
        <section className="cardContainer">
          <Buttons setData={this.setData} 
            favoriteCount={this.state.favoriteCount} 
            toggleDisplayFavorites={this.toggleDisplayFavorites} 
            toggleAddFavorites={this.toggleAddFavorites} />
          { this.toggleRender() }
        </section>
      );
    }
}

export default CardContainer;
