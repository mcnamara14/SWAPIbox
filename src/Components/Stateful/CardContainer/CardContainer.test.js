import React from 'react';
import ReactDOM from 'react-dom';
import CardContainer from './CardContainer';
import ScrollingText from '../ScrollingText/ScrollingText';
import Loading from '../../Stateless/Loading/Loading';
import { shallow, renderer } from 'enzyme';
import DataCleaner from '../../../DataCleaner/DataCleaner';
import { mockCleanedPeopleData } from '../../../DataCleaner/mockData';
jest.mock('../../../DataCleaner/DataCleaner', () => {
  return jest.fn().mockImplementation(() => {
    return {
      fetchData: jest.fn().mockImplementation(() => Promise.resolve({
        homeworld: "Tatooine",
        id: "people0",
        name: "Luke Skywalker",
        population: "200000",
        species: "Human"
      }))
    };
  });
});

describe('CardContainer', () => {
  let wrapper; 
  let dataCleaner;

  beforeEach((() => {
    wrapper = shallow(<CardContainer />)
  }))

  it('should match snapshot', () => {

    expect(wrapper).toMatchSnapshot();
  });

  describe('setData', () => {
    it('should call fetchData with correct argument', async () => {
      const filter = 'people';

      await wrapper.instance().setData(filter)

      expect(wrapper.instance().dataCleaner.fetchData).toHaveBeenCalledWith(filter)
    })
  })

  describe('findCard', () => {
    it('should call removeFavorite if a card is in favorites array already with selected card', () => {
      const id = 'vehicles0';
      const expected = { id: "vehicles0" };

      wrapper.setState({ 
        data: [{ id: "vehicles0" }],
        favorites: [{ id: "vehicles0" },{ id: "vehicles1"}]
      });

      wrapper.instance().removeFavorite =  jest.fn()
      wrapper.instance().findCard(id);

      expect(wrapper.instance().removeFavorite).toHaveBeenCalledWith(expected);
    })

    it('should call addFavorite if a card is in favorites array already with selected card', () => {
      const id = 'vehicles0';
      const expected = { id: "vehicles0" };

      wrapper.setState({ 
        data: [{ id: "vehicles0" }],
        favorites: [{ id: "vehicles1" },{ id: "vehicles2"}]
      });
      
      wrapper.instance().addFavorite =  jest.fn()
      wrapper.instance().findCard(id);

      expect(wrapper.instance().addFavorite).toHaveBeenCalledWith(expected);
    })
  })

  describe('removeFavorite', () => {
    it('should remove selected card from favorites array', () => {
      const mockSelectedCard = { id: "vehicles1" }
      const expected = [{ id: "vehicles2" }]

      wrapper.setState({ 
        favorites: [{ id: "vehicles1" },{ id: "vehicles2" }]
      });

      wrapper.instance().removeFavorite(mockSelectedCard);

      expect(wrapper.state().favorites).toEqual(expected)
    })

    it('should reduce favorite count by one when called', () =>{
      const mockSelectedCard = { id: "vehicles1" }

      wrapper.setState({ favoriteCount: 2 });

      wrapper.instance().removeFavorite(mockSelectedCard);

      expect(wrapper.state().favoriteCount).toEqual(1)
    })
  })

  describe('addFavorite', () => {
    it('should add selected card to favorites array', () => {
      const mockSelectedCard = { id: "vehicles2" }
      const expected = [{ id: "vehicles1" },{ id: "vehicles2" } ]

      wrapper.setState({ 
        favorites: [{ id: "vehicles1" }]
      });

      wrapper.instance().addFavorite(mockSelectedCard)

      expect(wrapper.state().favorites).toEqual(expected)
    })

    it('should add one to favorite count when called', () =>{
      const mockSelectedCard = { id: "vehicles1" }

      wrapper.setState({ favoriteCount: 1 });

      wrapper.instance().addFavorite(mockSelectedCard);

      expect(wrapper.state().favoriteCount).toEqual(2)
    })
  })

  describe('toggleAddFavorites', () => {
    it('should set render state to addFavorites when called', () => {
      const expected = 'addFavorites';

      wrapper.instance().toggleAddFavorites();
 
      expect(wrapper.state().renderState).toEqual(expected)
    })
  })


  describe('addFavorite', () => {
    it('should set data to favorites array when called', () => {
      const expected = [{ id: "vehicles1" }]

      wrapper.setState({ 
        data: [mockCleanedPeopleData],
        favorites: [{ id: "vehicles1" }]
      });

      wrapper.instance().toggleDisplayFavorites();

      expect(wrapper.state().data).toEqual(expected)
    })
  })

  describe('toggleRender', () => {
    it('should render scrolling text component to page when render state is crawl', () => {
      wrapper.setState({ 
        renderState: 'crawl'
      });

      wrapper.instance().toggleRender()

      expect(wrapper.find('ScrollingText').length).toEqual(1);
    })

    it.skip('should call getCards function when render state is cards', () => {
      wrapper.setState({ 
        renderState: 'cards'
      });

      wrapper.instance().getCards = jest.fn();
      wrapper.instance().toggleRender()

      expect(wrapper.instance().getCards).toHaveBeenCalled();
    })

    it('should render the Loading component when render state is loading', () => {
      wrapper.setState({ 
        renderState: 'loading'
      });

      wrapper.instance().toggleRender()

      expect(wrapper.find('Loading').length).toEqual(1);
    })
  })

})


