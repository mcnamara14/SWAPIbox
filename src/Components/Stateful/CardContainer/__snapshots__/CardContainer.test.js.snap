// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardContainer should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CardContainer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Buttons
          favoriteCount={0}
          setData={[Function]}
          toggleAddFavorites={[Function]}
          toggleDisplayFavorites={[Function]}
/>,
        <ScrollingText />,
      ],
      "className": "cardContainer",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "favoriteCount": 0,
          "setData": [Function],
          "toggleAddFavorites": [Function],
          "toggleDisplayFavorites": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Buttons
            favoriteCount={0}
            setData={[Function]}
            toggleAddFavorites={[Function]}
            toggleDisplayFavorites={[Function]}
/>,
          <ScrollingText />,
        ],
        "className": "cardContainer",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "favoriteCount": 0,
            "setData": [Function],
            "toggleAddFavorites": [Function],
            "toggleDisplayFavorites": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
